core/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java

```diff
@@ -0,0 +1,144 @@
+     */
+    //@Test
+    public void timingPBKDF2() {
+        // ignore the salt generation - check just matching....
+        PBKDF2PasswordEncoder encoder = new PBKDF2PasswordEncoder();
+        String encoded = encoder.encode("thisIsMyPassword1");
+
+        long start = System.nanoTime();
+        for (int i = 0; i < 10; i++) {
+            System.out.println(encoder.matches("thisIsMyPassword" + i, encoded));
+        }
+        long end = System.nanoTime();
+        long duration = end - start;
+        long duration_per_iteration = duration / 10;
+
+        Duration d = Duration.ofNanos(duration_per_iteration);
+        System.out.println("PBKDF2 took " + d.toNanos() + "ns");
+        System.out.println("PBKDF2 took " + d.toMillis() + "ms");
+        System.out.println("PBKDF2 took " + d.toSeconds() + "s");
+    }
+
+    /* This exists so that we can easily  check the complexity of how long this takes (ie is the number of iterations we 
+     * use correct for the state of CPUs).  
+     * We do not want to assert that the range < x and  > y  as that would make the test flaky on overloaded
+     * or slow hardware, so this is commented out but left for ease of running locally when desired.
+     */
+    //@Test
+    public void timingBcrypt() {
+        // ignore the salt generation - check just matching....
+        JBCryptEncoder encoder = new JBCryptEncoder();
+        String encoded = encoder.encode("thisIsMyPassword1");
+
+        long start = System.nanoTime();
+        for (int i = 0; i < 10; i++) {
+            System.out.println(encoder.matches("thisIsMyPassword" + i, encoded));
+        }
+        long end = System.nanoTime();
+        long duration = end - start;
+        long duration_per_iteration = duration / 10;
+
+        Duration d = Duration.ofNanos(duration_per_iteration);
+        System.out.println("BCrypt took " + d.toNanos() + "ns");
+        System.out.println("BCrypt took " + d.toMillis() + "ms");
+        System.out.println("BCrypt took " + d.toSeconds() + "s");
+    }
+
+    @Test
+    public void testPBKDF2RegExp() {
+        PBKDF2PasswordEncoder encoder = new PBKDF2PasswordEncoder();
+        String encoded = encoder.encode("thisIsMyPassword");
+        assertTrue(encoder.isHashValid(encoded));
+
+        // and a static one for other tests...
+        assertTrue(encoder.isHashValid(PBKDF2_HMAC_SHA512_ENCODED_PASSWORD));
+
+        assertFalse(encoder.isHashValid(
+                "$HMACSHA512:1000:fe899fcfcef4302ec3f0d36164efefdc$0781364eae9dac4ef1c4c3bf34c28e13965b46105fec0b6fcf4bae78e246fb5e51a1694fff19acac2dfb37b16055092644f3682c25beea9a7a286bf94e52f63b"),
+                "not enough iterations");
+        assertFalse(encoder.isHashValid(
+                "$HMACSHA512:500000:fe899fcfcef4302ec3f0d36164efefdc$0781364eae9dac4ef1c4c3bf34c28e13965b46105fec0b6fcf4bae78e246fb5e51a1694fff19acac2dfb37b16055092644f3682c25beea9a7a286bf94e52f63b"),
+                "too many iterations");
```

Comment:
Why is this a problem? (unreasonable number of iterations, maybe, but this is still the same order of magnitude).

Reply:
it does not match the Regular expression that this is specifically testing - it does not matter if we are over by 1, one thousand, or 1 million.

Reply:
Seems unnecessarily restrictive, and the test alone does not indicate why this is expected to fail. Just a nit though.